src/main/telemetry/ibus.c

```diff
@@ -0,0 +1,340 @@
+ *     0xA1       - bits 7-4 Command (1010 = request measurement)
+ *                  bits 3-0 Address (0001 = address 1)
+ *     0x5A, 0xFF - Checksum, 0xFFFF - (0x04 + 0xA1)
+ *
+ *
+ *                                                      /--------\
+ *                                I'm reading 0 volts < | Sensor |
+ *                [ 0x06 0xA1 0x00 0x00 0x5E 0xFF ]     \--------/
+ *
+ *     0x06       - Packet Length
+ *     0xA1       - bits 7-4 Command (1010 = request measurement)
+ *                  bits 3-0 Address (0001 = address 1)
+ *     0x00, 0x00 - The measurement
+ *     0x5E, 0xFF - Checksum, 0xFF - (0x06 + 0xA1 + 0x00 + 0x00)
+ *
+ *
+ * Due to the limited telemetry data types possible with ibus, we
+ * simply send everything which can be represented. Currently this
+ * is voltage and temperature.
+ *
+ */
+
+#define IBUS_UART_MODE     (MODE_RXTX)
+#define IBUS_BAUDRATE      (115200)
+#define IBUS_CYCLE_TIME_MS (8)
+
+#define IBUS_CHECKSUM_SIZE (2)
+
+#define IBUS_MIN_LEN       (2 + IBUS_CHECKSUM_SIZE)
+#define IBUS_MAX_TX_LEN    (6)
+#define IBUS_MAX_RX_LEN    (4)
+#define IBUS_RX_BUF_LEN    (IBUS_MAX_RX_LEN)
+
+#define IBUS_TEMPERATURE_OFFSET (0x0190)
+
+typedef uint8_t ibusAddress_t;
+
+typedef enum {
+    IBUS_COMMAND_DISCOVER_SENSOR      = 0x80,
+    IBUS_COMMAND_SENSOR_TYPE          = 0x90,
+    IBUS_COMMAND_MEASUREMENT          = 0xA0
+} ibusCommand_e;
+
+typedef enum {
+    IBUS_SENSOR_TYPE_INTERNAL_VOLTAGE = 0x00,
+    IBUS_SENSOR_TYPE_TEMPERATURE      = 0x01,
+    IBUS_SENSOR_TYPE_RPM              = 0x02,
+    IBUS_SENSOR_TYPE_EXTERNAL_VOLTAGE = 0x03
+} ibusSensorType_e;
+
+static const uint8_t SENSOR_ADDRESS_TYPE_LOOKUP[] = {
+    IBUS_SENSOR_TYPE_INTERNAL_VOLTAGE,  // Address 0, not usable since it is reserved for internal voltage
+    IBUS_SENSOR_TYPE_EXTERNAL_VOLTAGE,  // Address 1, VBAT
+    IBUS_SENSOR_TYPE_TEMPERATURE,       // Address 2, Gyro Temp
+    IBUS_SENSOR_TYPE_RPM                // Address 3, Throttle command
+};
+
+static serialPort_t *ibusSerialPort = NULL;
+static serialPortConfig_t *ibusSerialPortConfig;
+
+static telemetryConfig_t *telemetryConfig;
+static bool ibusTelemetryEnabled = false;
+static portSharing_e ibusPortSharing;
+
+static uint8_t ibusReceiveBuffer[IBUS_RX_BUF_LEN] = { 0x0 };
+
+static uint16_t calculateChecksum(uint8_t ibusPacket[static IBUS_CHECKSUM_SIZE], size_t packetLength) {
+    uint16_t checksum = 0xFFFF;
+    for (size_t i = 0; i < packetLength - IBUS_CHECKSUM_SIZE; i++) {
+        checksum -= ibusPacket[i];
+    }
+    return checksum;
+}
+
+static bool isChecksumOk(uint8_t ibusPacket[static IBUS_CHECKSUM_SIZE], size_t packetLength, uint16_t calcuatedChecksum) {
+    // Note that there's a byte order swap to little endian here
+    return (calcuatedChecksum >> 8) == ibusPacket[packetLength - 1]
+            && (calcuatedChecksum & 0xFF) == ibusPacket[packetLength - 2];
+}
+
+static void transmitIbusPacket(uint8_t ibusPacket[static IBUS_MIN_LEN], size_t packetLength) {
```

Comment:
Another possibility is to send checksum without storing it in buffer. Then it won't be necessary to allocate checksum bytes in packet .. but `packetLength` gets a bit confusing

static void transmitIbusPacket(uint8_t ibusPacket[static IBUS_MIN_LEN], size_t packetLength) {
    uint16_t checksum = calculateChecksum(ibusPacket, packetLength);
    for (size_t i = 0; i < packetLength - IBUS_CHECKSUM_SIZE; i++) {
        serialWrite(ibusSerialPort, ibusPacket[i]);
    }
    serialWrite(ibusSerialPort, (checksum & 0xFF));
   serialWrite(ibusSerialPort, (checksum >> 8));
}